File: iOS/PulsarLottie/Sources/PulsarLottie/Sampler.swift
import Foundation
import Pulsar

/// Reads a Pulsar ``PatternData`` into per-frame `RealtimeComposer` inputs.
///
/// Flattening the pattern once up front keeps the display-link step to plain arithmetic over
/// plain arrays. No UIKit/CoreHaptics dependencies here, so it is trivially unit-testable.

/// A single point on a continuous envelope.
struct EnvPoint {
    let time: Double
    let value: Float
}

/// A discrete transient.
struct DiscEvent {
    let time: Double
    let amplitude: Float
    let frequency: Float
}

/// A `PatternData` flattened into plain arrays ready for sampling.
struct SampledPattern {
    let amplitude: [EnvPoint]
    let frequency: [EnvPoint]
    let discrete: [DiscEvent]

    var hasContinuous: Bool { !amplitude.isEmpty && !frequency.isEmpty }
}

/// Flatten a `PatternData` into a ``SampledPattern``.
func sampledPattern(from pattern: PatternData) -> SampledPattern {
    SampledPattern(
        amplitude: pattern.continuousPattern.amplitude.map { EnvPoint(time: $0.time, value: $0.value) },
        frequency: pattern.continuousPattern.frequency.map { EnvPoint(time: $0.time, value: $0.value) },
        discrete: pattern.discretePattern.map { DiscEvent(time: $0.time, amplitude: $0.amplitude, frequency: $0.frequency) }
    )
}

/// Linear-interpolate a breakpoint envelope at time `t` (ms).
func sampleEnvelope(_ points: [EnvPoint], _ t: Double) -> Float {
    let n = points.count
    if n == 0 { return 0 }
    let first = points[0]
    if t <= first.time { return first.value }
    let last = points[n - 1]
    if t >= last.time { return last.value }
    for i in 1..<n {
        let b = points[i]
        if t <= b.time {
            let a = points[i - 1]
            let span = b.time - a.time
            if span <= 0 { return b.value }
            let k = Float((t - a.time) / span)
            return a.value + (b.value - a.value) * k
        }
    }
    return last.value
}

/// Total pattern length in ms — the largest timestamp across all channels.
func patternDurationMs(_ pattern: SampledPattern) -> Double {
    var maxT = 0.0
    for e in pattern.discrete where e.time > maxT { maxT = e.time }
    for e in pattern.amplitude where e.time > maxT { maxT = e.time }
    for e in pattern.frequency where e.time > maxT { maxT = e.time }
    return maxT
}

/// Clamp `v` into `[0, 1]`.
func clamp01(_ v: Float) -> Float {
    if v < 0 { return 0 }
    if v > 1 { return 1 }
    return v
}

File: react-native/PulsarLottie/src/internal/sampler.ts
import type { Pattern } from 'react-native-pulsar';

/**
 * Worklet-safe helpers that turn a Pulsar {@link Pattern} into per-frame
 * `RealtimeComposer` inputs. Everything here runs on the UI thread inside a
 * Reanimated frame callback, so it uses only plain arithmetic over plain
 * arrays (no closures over React state).
 */

export interface EnvelopePoint {
  time: number;
  value: number;
}

/** Linear-interpolate a breakpoint envelope (`{time,value}[]`) at time `t` (ms). */
export function sampleEnvelope(points: readonly EnvelopePoint[], t: number): number {
  'worklet';
  const n = points.length;
  if (n === 0) {
    return 0;
  }
  const first = points[0]!;
  if (t <= first.time) {
    return first.value;
  }
  const last = points[n - 1]!;
  if (t >= last.time) {
    return last.value;
  }
  for (let i = 1; i < n; i++) {
    const b = points[i]!;
    if (t <= b.time) {
      const a = points[i - 1]!;
      const span = b.time - a.time;
      if (span <= 0) {
        return b.value;
      }
      const k = (t - a.time) / span;
      return a.value + (b.value - a.value) * k;
    }
  }
  return last.value;
}

/** Clamp `v` into `[lo, hi]`. */
export function clamp(v: number, lo: number, hi: number): number {
  'worklet';
  return v < lo ? lo : v > hi ? hi : v;
}

/** Total pattern length in ms — the largest timestamp across all channels. */
export function patternDurationMs(pattern: Pattern): number {
  let max = 0;
  for (let i = 0; i < pattern.discretePattern.length; i++) {
    max = Math.max(max, pattern.discretePattern[i]!.time);
  }
  for (let i = 0; i < pattern.continuousPattern.amplitude.length; i++) {
    max = Math.max(max, pattern.continuousPattern.amplitude[i]!.time);
  }
  for (let i = 0; i < pattern.continuousPattern.frequency.length; i++) {
    max = Math.max(max, pattern.continuousPattern.frequency[i]!.time);
  }
  return max;
}

/** Derive the animation duration (ms) from an inline Lottie JSON source, if present. */
export function lottieDurationMs(source: unknown): number | undefined {
  if (source && typeof source === 'object') {
    const doc = source as { fr?: number; ip?: number; op?: number };
    if (typeof doc.fr === 'number' && typeof doc.op === 'number') {
      const fr = doc.fr || 30;
      const ip = typeof doc.ip === 'number' ? doc.ip : 0;
      const op = doc.op;
      if (fr > 0 && op > ip) {
        return ((op - ip) / fr) * 1000;
      }
    }
  }
  return undefined;
}
